OpenRA.Mods.Common/Traits/Sound/AmbientSound.cs

```diff
@@ -19,56 +19,87 @@ class AmbientSoundInfo : ConditionalTraitInfo
 		[FieldLoader.Require]
 		public readonly string SoundFile = null;
 
-		[Desc("Interval between playing the sound (in ticks).")]
-		public readonly int Interval = 0;
+		[Desc("Initial delay (in ticks) before playing the sound for the first time.",
+			"Two values indicate a random delay range.")]
+		public readonly int[] Delay = { 0 };
+
+		[Desc("Interval between playing the sound (in ticks).",
+			"Two values indicate a random delay range.")]
+		public readonly int[] Interval = { 0 };
 
 		public override object Create(ActorInitializer init) { return new AmbientSound(init.Self, this); }
 	}
 
-	class AmbientSound : ConditionalTrait<AmbientSoundInfo>, ITick
+	class AmbientSound : ConditionalTrait<AmbientSoundInfo>, ITick, INotifyRemovedFromWorld
 	{
+		readonly bool loop;
 		ISound currentSound;
-		bool wasDisabled = true;
-		int interval;
+		WPos cachedPosition;
+		int delay;
 
 		public AmbientSound(Actor self, AmbientSoundInfo info)
 			: base(info)
 		{
-			interval = info.Interval;
+			delay = RandomDelay(self.World, info.Delay);
+			loop = Info.Interval.Length == 0 || (Info.Interval.Length == 1 && Info.Interval[0] == 0);
 		}
 
-		public void Tick(Actor self)
+		void ITick.Tick(Actor self)
 		{
 			if (IsTraitDisabled)
-			{
-				Game.Sound.StopSound(currentSound);
-				currentSound = null;
-				wasDisabled = true;
 				return;
+
+			var pos = self.CenterPosition;
+			if (currentSound != null && pos != cachedPosition)
+			{
+				currentSound.SetPosition(pos);
+				cachedPosition = pos;
 			}
 
-			if (wasDisabled && Info.Interval <= 0)
+			if (delay >= 0 && --delay < 0)
```

Comment:
Just a reminder of our discussion on IRC: Please pull the `< 0` check out into an own check.